Add corfu quick and corfu indexed

Luis-Henriquez-Perez opened the pull request:
This PR is for adding `corfu-quick` and `corfu-indexed`.

I include both of them together to address what you said in number #110: "Ideally I'd like to merge these two together, such that we see if changes to corfu.el itself are necessary".

Also I will reiterate the summary I wrote in #110:

corfu-quick is almost done. A quick summary:

corfu-quick--format-candidates - works.

corfu-quick--letters - should work, but I need to test indices at end of string length. I'll finish testing today.

corfu-quick--read - (corfu--candidates-popup (point)) stopped working for me. The popup wouldn't show. I need to try it again and investigate it by going to an earlier commit where it did work. Also maybe I need to deal with the case the user passes in out of range characters (or maybe corfu-quick-jump does this).

corfu-quick-jump - works.

I added corfu-quick-exit for being able to referesh the popup without the corfu-quick letters in case the user changes their mind. Initially, I wrote it for my own testing but it may be useful all on its own.

minad commented on corfu-indexed.el:
```diff
@@ -0,0 +1,87 @@
+;;; corfu.el --- Completion Overlay Region FUnction -*- lexical-binding: t -*-
```
name

minad replied:
the file should live in a subdirectory extension/ like in vertico.

minad commented on corfu-indexed.el:
```diff
@@ -0,0 +1,87 @@
+;;; corfu.el --- Completion Overlay Region FUnction -*- lexical-binding: t -*-
+
+;; Copyright (C) 2021  Free Software Foundation, Inc.
```
date

minad commented on corfu-indexed.el:
```diff
@@ -0,0 +1,87 @@
+;; Author: Daniel Mendler <mail@daniel-mendler.de>
+;; Maintainer: Daniel Mendler <mail@daniel-mendler.de>
+;; Created: 2021
+;; Version: 0.17
```
version

minad commented on corfu-indexed.el:
```diff
@@ -0,0 +1,87 @@
+;; Maintainer: Daniel Mendler <mail@daniel-mendler.de>
+;; Created: 2021
+;; Version: 0.17
+;; Package-Requires: ((emacs "27.1"))
```
should require corfu

Luis-Henriquez-Perez replied:
Required corfu.

minad commented on corfu-indexed.el:
```diff
@@ -0,0 +1,87 @@
+;;; Code:
+
+(require 'corfu)
+(require 'pcase)
```
pcase require is not needed

Luis-Henriquez-Perez replied:
Removed pcase require.

minad commented on corfu-indexed.el:
```diff
@@ -0,0 +1,87 @@
+(defvar corfu-indexed--commands
+  '(corfu-insert corfu-complete))
+
+(defvar-local corfu-indexed--max 0)
```
Use defvar here.

Luis-Henriquez-Perez replied:
Done.

minad commented on corfu-indexed.el:
```diff
@@ -0,0 +1,87 @@
+
+(defvar-local corfu-indexed--max 0)
+
+(defun corfu-indexed--format-candidate (orig cands)
```
Name this format-candidates since it formats multiple candidates.

Luis-Henriquez-Perez replied:
Done.

minad commented on corfu-indexed.el:
```diff
@@ -0,0 +1,87 @@
+        (if (or (< corfu--index 0)
+                (> corfu--index corfu-indexed--max)
+                (= corfu--total 0))
+            (minibuffer-message "Out of range")
```
message instead of minibuffer-message

Luis-Henriquez-Perez replied:
Done.

minad commented on corfu-quick.el:
```diff
@@ -0,0 +1,135 @@
+;; Created: 2021
+;; Version: 0.17
+;; Package-Requires: ((emacs "27.1"))
+;; Homepage: https://github.com/minad/corfu
```
Same updates needed here. You can add yourself as author.

Luis-Henriquez-Perez replied:
Updated year and version. Also added myself as author.

minad commented on corfu-quick.el:
```diff
@@ -0,0 +1,135 @@
+
+;;; Commentary:
+
+;; This package is a corfu extension, which provides quick keys. Quick keys are
```
Corfu in upper case, since it is a name.

Luis-Henriquez-Perez replied:
Ok. Since "quick keys" is a term, as opposed to just describing keys as "quick" should we add a hyphen in between: "quick-keys"?

Luis-Henriquez-Perez replied:
I'll leave it as it is.

minad commented on corfu-quick.el:
```diff
@@ -0,0 +1,135 @@
+  (require 'cl-lib)
+  (require 'subr-x)
+  (require 'pcase)
+  (require 'seq))
```
1. require corfu at runtime
2. remove the rest

Luis-Henriquez-Perez replied:
Done.

minad commented on corfu-quick.el:
```diff
@@ -0,0 +1,135 @@
+;;;###autoload
+(defun corfu-quick-exit ()
+  "Remove quick keys from popup."
+  (corfu--candidates-popup (point)))
```
Remove this function. Why is it needed?

Luis-Henriquez-Perez replied:
> Remove this function. Why is it needed?

Removed.

I put it there so the user could quit out of the quick keys if they change their mind. I'm not 100% convinced it's necessary myself so I'll remove it for now.

minad commented:
I added a few comments pointing out minor superficial issues. Please take a lot of care for the details, every single word counts. :)

But besides that I don't see major issues, I will try this soon and then I can tell you more.

Luis-Henriquez-Perez commented:
> I added a few comments pointing out minor superficial issues.

Ok. I will fix these.

> I will try this soon and then I can tell you more.

Ok. Please try it out after the `Fix corfu-quick--read` commit. I fixed many problems there.

minad commented:
> Ok. Please try it out after the Fix corfu-quick--read commit. I fixed many problems there.

Please take your time and polish everything. I will try your patch in due time!

minad commented:
@Luis-Henriquez-Perez Now that you've finally got the assignment :smile:, how is the status here? Is this ready to review? There are still a few open nitpitcks.

Luis-Henriquez-Perez commented on corfu-quick.el:
```diff
@@ -0,0 +1,135 @@
+  "Read quick keys and return index of candidate specified by quick keys."
+  (cl-letf* ((old-fn (symbol-function #'corfu--format-candidates))
+	     (new-fn (apply-partially #'corfu-quick--format-candidates old-fn))
+	     ((symbol-function #'corfu--format-candidates) new-fn))
```
I didn't change this because I'm not sure how the indentation should be. `aggressive-indent` isn't complaining for me.

Luis-Henriquez-Perez commented:
@minad I think this is ready to review. 

I resolved the all the nitpicks except for the indentation of `cl-letf*` which I am not sure about.

Also, I rebased to the main branch.

Finally, I tested corfu quick by simply evaluating the corfu-quick file and binding `corfu-quick-insert` to a key and trying it out. I did the same thing with `corfu-indexed`. However, I had trouble entering the prefix arguments, but I suspect my own config is the culprit.
